checker/errors.go

```diff
@@ -44,41 +45,68 @@ func (errors *TypeErrors) fieldDoesNotSupportPresenceCheck(l common.Location, fi
 }
 
-func (errors *TypeErrors) aggregateTypeMismatch(l common.Location, aggregate types.Type, member types.Type) {
+func (errors *TypeErrors) aggregateTypeMismatch(l common.Location, aggregate *checked.Type, member *checked.Type) {
 	errors.ReportError(
 		l,
 		"type '%s' does not match previous type '%s' in aggregate. Use 'dyn(x)' to make the aggregate dynamic.",
-		member,
-		aggregate)
+		types.FormatType(member),
+		types.FormatType(aggregate))
 }
 
-func (errors *TypeErrors) notAType(l common.Location, t types.Type) {
-	errors.ReportError(l, "'%s(%v)' is not a type", t.String(), t.Kind())
+func (errors *TypeErrors) notAType(l common.Location, t *checked.Type) {
+	errors.ReportError(l, "'%s(%v)' is not a type", types.FormatType(t), t)
 }
 
-func (errors *TypeErrors) notAMessageType(l common.Location, t types.Type) {
-	errors.ReportError(l, "'%s' is not a message type", t.String())
+func (errors *TypeErrors) notAMessageType(l common.Location, t *checked.Type) {
+	errors.ReportError(l, "'%s' is not a message type", types.FormatType(t))
 }
 
-func (errors *TypeErrors) fieldTypeMismatch(l common.Location, name string, field types.Type, value types.Type) {
+func (errors *TypeErrors) fieldTypeMismatch(l common.Location, name string, field *checked.Type, value *checked.Type) {
 	errors.ReportError(l, "expected type of field '%s' is '%s' but provided type is '%s'",
-		name, field.String(), value.String())
+		name, types.FormatType(field), types.FormatType(value))
 }
 
 func (errors *TypeErrors) unexpectedFailedResolution(l common.Location, typeName string) {
 	errors.ReportError(l, "[internal] unexpected failed resolution of '%s'", typeName)
 }
 
-func (errors *TypeErrors) notAComprehensionRange(l common.Location, t types.Type) {
+func (errors *TypeErrors) notAComprehensionRange(l common.Location, t *checked.Type) {
 	errors.ReportError(l, "expression of type '%s' cannot be range of a comprehension (must be list, map, or dynamic)",
-		t.String())
+		types.FormatType(t))
 }
 
-func (errors *TypeErrors) typeMismatch(l common.Location, expected types.Type, actual types.Type) {
-	errors.ReportError(l, "expected type '%s' but found '%s'", expected, actual)
+func (errors *TypeErrors) typeMismatch(l common.Location, expected *checked.Type, actual *checked.Type) {
+	errors.ReportError(l, "expected type '%s' but found '%s'",
+		types.FormatType(expected), types.FormatType(actual))
+}
+
+func formatFunction(resultType *checked.Type, argTypes []*checked.Type, isInstance bool) string {
+	result := ""
+	if isInstance {
+		target := argTypes[0]
+		argTypes = argTypes[1:]
+
+		result += types.FormatType(target)
+		result += "."
+	}
+
+	result += "("
+	for i, arg := range argTypes {
+		if i > 0 {
+			result += ", "
+		}
+		result += types.FormatType(arg)
+	}
+	result += ")"
+	if resultType != nil {
+		result += " -> "
+		result += types.FormatType(resultType)
+	}
+
+	return result
```

Comment:
If you care about performance you might want to check out http://herman.asia/efficient-string-concatenation-in-go which suggests that something like the following would be much faster:

```go
func formatFunction(resultType *checked.Type, argTypes []*checked.Type, isInstance bool) string {
	var result bytes.Buffer
	if isInstance {
		target := argTypes[0]
		argTypes = argTypes[1:]

		result.WriteString(types.FormatType(target))
		result.WriteString(".")
	}

	result.WriteString("(")
	for i, arg := range argTypes {
		if i > 0 {
			result.WriteString(", ")
		}
		result.WriteString(types.FormatType(arg))
	}
	result.WriteString(")")
	if resultType != nil {
		result.WriteString(" -> ")
		result.WriteString(types.FormatType(resultType))
	}

	return result.String()
}
```

Reply:
Thanks for the tip. I don't particularly care about performance here, but this is good to know.